File: sdk-aws-python/src/getApiTokenEncrypted.py
# Import necessary libraries
from dotenv import load_dotenv
import os
import json
import time
import boto3
import base64
import requests
import logging
from urllib.parse import urlencode
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

from cryptography.hazmat.primitives.padding import PKCS7
from cryptography.hazmat.backends import default_backend

# Load environment variables from .env file
load_dotenv()

# Configure the AWS SDK to use LocalStack for testing
is_local = os.getenv("IS_LOCAL")
endpoint_url = "http://localhost:4566" if is_local else None
region_name = "us-east-1"

# Initialize AWS services clients
dynamodb = boto3.resource(
    "dynamodb", region_name=region_name, endpoint_url=endpoint_url
)
table = dynamodb.Table("Tokens")
secretsmanager = boto3.client(
    "secretsmanager", region_name=region_name, endpoint_url=endpoint_url
)


# Function to retrieve secrets from AWS Secrets Manager
def get_secrets():
    secret_name = os.getenv("SECRET_NAME")
    if not secret_name:
        raise ValueError("SECRET_NAME environment variable is not set")
    try:
        response = secretsmanager.get_secret_value(SecretId=secret_name)
        if "SecretString" in response:
            return json.loads(response["SecretString"])
        else:
            return base64.b64decode(response["SecretBinary"]).decode("utf-8")
    except Exception as e:
        print(f"Failed to retrieve secrets: {e}")
        raise


# Function to generate encryption key
def generate_encryption_key(consumer_key, consumer_secret):
    logging.info("Generating encryption key.")

    try:
        if not consumer_key or not consumer_secret:
            logging.error("consumerKey or consumerSecret is undefined.")
            raise ValueError("consumerKey or consumerSecret is undefined")

        digest = hashes.Hash(hashes.SHA256(), backend=default_backend())
        digest.update(consumer_key.encode("utf-8") + consumer_secret.encode("utf-8"))
        encryption_key = digest.finalize()[:32]
        logging.info("Encryption key generated successfully.")
        return encryption_key

    except Exception as e:
        logging.error(f"An error occurred during encryption key generation: {e}")
        raise


# Function to encrypt data
def encrypt(data, key):
    try:
        iv = os.urandom(16)
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        encryptor = cipher.encryptor()
        # Add PKCS7 padding to the data
        padder = PKCS7(128).padder()  # 128-bit block size for AES
        padded_data = padder.update(data.encode("utf-8")) + padder.finalize()
        encrypted_data = encryptor.update(padded_data) + encryptor.finalize()
        b64_encoded_encrypted_data = base64.b64encode(iv + encrypted_data).decode(
            "utf-8"
        )
        return b64_encoded_encrypted_data
    except Exception as e:
        print(f"An error occurred during encryption: {e}")

        raise


# Function to decrypt data
def decrypt(data, key):
    try:
        data = base64.b64decode(data)
        iv = data[:16]
        encrypted_data = data[16:]
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        decryptor = cipher.decryptor()
        padded_data = decryptor.update(encrypted_data) + decryptor.finalize()

        # Remove PKCS7 padding
        unpadder = PKCS7(128).unpadder()  # 128-bit block size for AES
        decrypted_data = unpadder.update(padded_data) + unpadder.finalize()

        try:
            return json.loads(decrypted_data.decode("utf-8"))
        except UnicodeDecodeError:
            print(f"Cannot decode decrypted data as UTF-8: {decrypted_data}")
            raise ValueError("Decrypted data is not a valid UTF-8 string")
    except Exception as e:
        print(f"An error occurred during decryption: {e}")

        raise


# Function to save token to DynamoDB
def save_token_to_dynamodb(token, key):
    encrypted_token = encrypt(json.dumps(token), key)
    table.put_item(Item={"Id": "apiToken1", "token": encrypted_token})


# Function to load token from DynamoDB
def load_token_from_dynamodb(key):
    response = table.get_item(Key={"Id": "apiToken1"})
    if "Item" in response:
        decrypted_data = decrypt(response["Item"]["token"], key)
        token = decrypted_data
        if time.time() < token["expiry"]:
            return token
    return None


# Function to fetch token from the server
def fetch_token(url, headers, params):
    response = requests.post(url, headers=headers, data=params)
    response.raise_for_status()
    return response.json()


# Function to retry fetching the token
def retry_fetch_token(retries, interval, fetch_token_func):
    for i in range(retries):
        try:
            return fetch_token_func()
        except Exception as e:
            print(f"Attempt {i+1} failed to refresh token: {e}")
            if i < retries - 1:
                print(f"Retrying in {interval / 1000} seconds...")
                time.sleep(interval / 1000)
            else:
                print(
                    "All retries failed. Please check your connection and credentials."
                )
                raise


# Function to refresh the token
def refresh_token(retries=2, interval=3000, should_retry=True):
    try:
        secrets = get_secrets()
        consumer_key = secrets["consumerKey"]
        consumer_secret = secrets["consumerSecret"]
        url = os.getenv("URL")
        scope = os.getenv("SCOPE")
        credentials = base64.b64encode(
            f"{consumer_key}:{consumer_secret}".encode("utf-8")
        ).decode("utf-8")
        headers = {
            "Authorization": f"Basic {credentials}",
            "Content-Type": "application/x-www-form-urlencoded",
        }
        params = urlencode({"grant_type": "client_credentials", "scope": scope})

        def fetch_token_func():
            return fetch_token(url, headers, params)

        encryption_key = generate_encryption_key(consumer_key, consumer_secret)
        token = load_token_from_dynamodb(encryption_key)
        if not token:
            if should_retry:
                token = retry_fetch_token(retries, interval, fetch_token_func)
            else:
                token = fetch_token_func()
            token["expiry"] = time.time() + token["expires_in"]
            save_token_to_dynamodb(token, encryption_key)
        return token

    except Exception as e:
        print(f"An error occurred during token refreshing: {e}")

        raise

File: sdk-aws-lambda-nodejs/src/getApiTokenEncrypted.js
/* eslint-disable no-undef */
require("dotenv").config();
const crypto = require("crypto");
// eslint-disable-next-line import/no-extraneous-dependencies
const AWS = require("aws-sdk");
const fetch = require("node-fetch");
// when using LocalStack for testing, comment the following line:
const dynamoDB = new AWS.DynamoDB.DocumentClient();
// if using multiple environments, you can customize DynamoDB table name and token ID through environment variables
const TABLE_NAME = process.env.TABLE_NAME || "Tokens";
const TOKEN_ID = process.env.TOKEN_ID || "apiToken";

/**
 * * Configure the AWS SDK to use LocalStack by uncommenting the following lines:
 * const isLocal = process.env.IS_LOCAL;
 * AWS.config.update({
 * region: "us-east-1",
 * endpoint: isLocal ? "http://localhost:4566" : undefined,
 * });
 * const dynamoDB = new AWS.DynamoDB.DocumentClient({
 * region: "us-east-1",
 *  endpoint: isLocal ? "http://localhost:4566" : undefined,
 * });
 */
const ALGORITHM = "aes-256-cbc";

/**
 * Retrieves secrets from AWS Secrets Manager.
 * @returns {Promise<Object>} The secrets.
 * @throws {Error} If the SECRET_NAME name variable is not set or secrets could not be retrieved.
 * @async
 */
async function getSecrets() {
  const secretsManager = new AWS.SecretsManager();
  const secretName = process.env.SECRET_NAME;
  if (!secretName) {
    throw new Error("SECRET_NAME environment variable is not set");
  }
  try {
    const data = await secretsManager
      .getSecretValue({ SecretId: secretName })
      .promise();
    if (data.SecretString) {
      return JSON.parse(data.SecretString);
    }
    // If the secret is binary, convert it to a string
    const buff = Buffer.from(data.SecretBinary, "base64");
    return buff.toString("ascii");
  } catch (error) {
    console.error("Failed to retrieve secrets:", error);
    throw error;
  }
}

/**
 * Generates an encryption key using the provided consumer key and secret.
 * @param {string} consumerKey - The consumer key.
 * @param {string} consumerSecret - The consumer secret.
 * @returns {Buffer} The generated encryption key.
 * @throws {Error} If the consumer key or secret is undefined.
 */
function generateEncryptionKey(consumerKey, consumerSecret) {
  if (!consumerKey || !consumerSecret) {
    throw new Error("consumerKey or consumerSecret is undefined");
  }
  const hash = crypto.createHash("sha256");
  const key = hash.digest().subarray(0, 32);
  return key;
}

/**
 * Encrypts the given data using the provided key.
 * @param {string} data - The data to encrypt.
 * @param {Buffer} key - The encryption key.
 * @returns {string} The encrypted data in hex format.
 * @throws {Error} If there's an error during encryption.
 */
function encrypt(data, key) {
  try {
    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv(ALGORITHM, key, iv);
    const encryptedData = Buffer.concat([cipher.update(data), cipher.final()]);
    return `${iv.toString("hex")}:${encryptedData.toString("hex")}`;
  } catch (error) {
    console.error("Error during encryption:", error);
    throw new Error(`Failed to encrypt data: ${error.message}`);
  }
}

/**
 * Decrypts the provided data using the provided key.
 * @param {string} data - The data to decrypt.
 * @param {Buffer} key - The decryption key.
 * @returns {string} The decrypted data.
 * @throws {Error} If there is an error while decrypting the data.
 */
function decrypt(data, key) {
  try {
    const [iv, encryptedData] = data
      .split(":")
      .map((part) => Buffer.from(part, "hex"));
    const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);
    const decryptedData = Buffer.concat([
      decipher.update(encryptedData),
      decipher.final(),
    ]);
    return decryptedData.toString();
  } catch (error) {
    console.error("Error during decryption:", error);
    return null;
  }
}

/**
 * Saves the provided token to a DynamoDB table. The token is encrypted using the provided key before being saved.
 * The table name and token ID can be customized through environment variables.
 * @param {Object} token - The token to save.
 * @param {Buffer} key - The encryption key.
 * @throws {Error} If there is an error while saving the token to DynamoDB.
 * @async
 */

async function saveTokenToDynamoDB(token, key) {
  try {
    const encryptedToken = encrypt(JSON.stringify(token), key);
    const params = {
      TableName: TABLE_NAME,
      Item: {
        Id: TOKEN_ID,
        token: encryptedToken,
      },
    };
    await dynamoDB.put(params).promise();
  } catch (error) {
    console.error(`Failed to save token to DynamoDB: ${error}`);
    throw error;
  }
}

/**
 * Decrypts the token from DynamoDB using the provided key.
 * @param {string} key - The encryption key.
 * @returns {Promise<string|null>} The decrypted data or null if not found.
 * @throws {Error} If there's an error during decryption or fetching from DynamoDB.
 * @async
 */


async function getDecryptedDataFromDynamoDB(key) {
  try {
    const params = {
      TableName: TABLE_NAME,
      Key: {
        Id: TOKEN_ID,
      },
    };
    const data = await dynamoDB.get(params).promise();
    if (data.Item) {
      return decrypt(data.Item.token, key);
    }
  } catch (error) {
    console.error("Error fetching from DynamoDB:", error);
    throw new Error(`Failed to fetch token from DynamoDB: ${error.message}`);
  }
  return null;
}

/**
 * Parses the decrypted data into a token object.
 * @param {string} decryptedData - The decrypted data.
 * @returns {Object|null} The token object or null if invalid or expired.
 * @throws {Error} If there's an error during parsing.
 */
function parseToken(decryptedData) {
  if (decryptedData) {
    try {
      const token = JSON.parse(decryptedData);
      if (Date.now() < token.expiry) {
        return token;
      }
    } catch (error) {
      console.error("Error parsing token:", error);
      throw new Error(`Failed to parse token: ${error.message}`);
    }
  }
  return null;
}

/**
 * Loads the token from DynamoDB, decrypts it, and checks its validity.
 * @param {string} key - The encryption key.
 * @returns {Promise<Object|null>} The token object or null if not found or invalid.
 * @throws {Error} If there's an error during the process.
 * @async
 */
async function loadTokenFromDynamoDB(key) {
  try {
    const decryptedData = await getDecryptedDataFromDynamoDB(key);
    return parseToken(decryptedData);
  } catch (error) {
    error.message = `Failed to load token from DynamoDB: ${error.message}`;
    throw error;
  }
}

/**
 * Fetches a token from the provided URL with the provided headers and body parameters.
 * @param {string} url - The URL to fetch from.
 * @param {Object} headers - The headers for the request.
 * @param {URLSearchParams} params - The body parameters for the request.
 * @returns {Promise<Object>} The JSON response from the fetch request.
 * @throws {Error} If the fetch request fails.
 */
async function fetchToken(url, headers, params) {
  const response = await fetch(url, {
    method: "POST",
    headers,
    body: params,
  });
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }
  return response.json();
}

/**
 * Fetches a token from the provided URL with the provided headers and body parameters.
 * @param {string} url - The URL to fetch from.
 * @param {Object} headers - The headers for the request.
 * @param {URLSearchParams} params - The body parameters for the request.
 * @returns {Promise<Object>} The JSON response from the fetch request.
 * @throws {Error} If the fetch request fails.
 * @async
 */
async function retryFetchToken(retries, interval, fetchTokenFunc, attempt = 0) {
  try {
    const result = await fetchTokenFunc();
    return result;
  } catch (error) {
    console.error(`Attempt ${attempt + 1} failed to refresh token: ${error}`);
    if (attempt < retries - 1) {
      console.log(`Retrying in ${interval / 1000} seconds...`);
      // eslint-disable-next-line no-promise-executor-return
      await new Promise((resolve) => setTimeout(resolve, interval)); // No need to return anything here
      return retryFetchToken(retries, interval, fetchTokenFunc, attempt + 1);
    }
    console.error(
      "All retries failed. Please check your connection and credentials.",
    );
    throw error;
  }
}

/**
 * Main function of the module. Attempts to refresh the token. If `shouldRetry` is true, it will use `retryFetchToken` to fetch the token, otherwise it will use `fetchToken`. If the token fetch is successful, it will save the token to a file and return it.
 * @param {number} [retries=process.env.RETRIES || 2] - The number of retries.
 * @param {number} [interval=process.env.INTERVAL || 3000] - The interval between retries in milliseconds.
 * @param {boolean} [shouldRetry=true] - Whether to retry fetching the token if the fetch fails.
 * @returns {Promise<Object>} The fetched token.
 * @returns {Promise<Object>} The API token.
 * @throws Will throw an error if loading the token from DynamoDB fails.
 * @async
 */
async function refreshToken(
  retries = process.env.RETRIES || 2,
  interval = process.env.INTERVAL || 3000,
  shouldRetry = true,
) {
  const secrets = await getSecrets();
  const { consumerKey } = secrets;
  const { consumerSecret } = secrets;
  const url = process.env.URL;
  const scope = process.env.SCOPE;

  const credentials = Buffer.from(`${consumerKey}:${consumerSecret}`).toString(
    "base64",
  );

  const headers = {
    Authorization: `Basic ${credentials}`,
    "Content-Type": "application/x-www-form-urlencoded",
  };

  const params = new URLSearchParams();
  params.append("grant_type", "client_credentials");
  params.append("scope", scope);
  const fetchTokenFunc = () => fetchToken(url, headers, params);

  const encryptionKey = generateEncryptionKey(consumerKey, consumerSecret);

  let token;
  try {
    token = await loadTokenFromDynamoDB(encryptionKey);
  } catch (error) {
    console.error("Error during loadTokenFromDynamoDB:", error);
  }

  if (!token) {
    if (shouldRetry) {
      token = await retryFetchToken(retries, interval, fetchTokenFunc);
    } else {
      token = await fetchTokenFunc();
    }
    token.expiry = Date.now() + token.expires_in * 1000;
    await saveTokenToDynamoDB(token, encryptionKey);
  }

  return token;
}

module.exports = {
  refreshToken,
  getSecrets,
  saveTokenToDynamoDB,
  loadTokenFromDynamoDB,
  fetchToken,
  retryFetchToken,
  encrypt,
  decrypt,
  generateEncryptionKey,
};
